gpkit/tests/t_examples.py

```diff
@@ -95,42 +96,34 @@ def __exit__(self, *args):
         gpkit.settings["installed_solvers"] = SOLVERS
 
 
-class NullFile(object):
-    def write(self, string):
-        pass
-
-    def close(self):
-        pass
-
+def logged_example_testcase(name,
+                            imported=IMPORTED_EXAMPLES,
+                            exampledir=EXAMPLE_DIR):
+    """Returns a method for attaching to a unittest.TestCase that imports
+    or reloads module 'name' and stores in imported[name].
+    Runs top-level code, which is typically a docs example, in the process.
 
-class StdoutCaptured(object):
-    def __init__(self, logfilename=None):
-        self.logfilename = logfilename
-
-    def __enter__(self):
-        self.original_stdout = sys.stdout
-        if self.logfilename:
-            filepath = EXAMPLE_DIR+os.sep+"%s_output.txt" % self.logfilename
-            logfile = open(filepath, "w")
-        else:
-            logfile = NullFile()
-        sys.stdout = logfile
-
-    def __exit__(self, *args):
-        sys.stdout.close()
-        sys.stdout = self.original_stdout
+    Returns a method.
+    """
+    def test(self):
+        filepath = ("".join([exampledir,
+                             os.sep,
+                             "%s_output.txt" % name])
+                    if name not in imported else None)
+        with StdoutCaptured(logfilepath=filepath):
+            if name not in imported:
+                imported[name] = importlib.import_module(name)
+            else:
+                reload(imported[name])
+        getattr(self, name)(imported[name])
```

Comment:
@bqpd: the lione above is surprisingly confusing. I know the tests get replaced with methods that take `example` as an argument, as beautifully detailed in the `TestExamples` docstring. But I'm confused because those methods have names like `test_water_tank_cvxopt`, whereas here `name` has values like `water_tank`.

Does the `getattr` above look up `self.water_tank` which has been moved by `setattr(TestExamples, name, old_test) # move to a non-test fn` below?

Do you have any ideas for touching up the docstring above and the code so it's a little more clear what's going on. This took me awhile to understand when trying to modify it for external use.